File: MetaGeta/MetaGeta.DataStore/Settings/SettingInfo.vb
Imports System.Reflection
Imports System.Runtime.CompilerServices

Public Class SettingInfo
    Implements INotifyPropertyChanged

    Private ReadOnly m_TargetObject As Object
    Private ReadOnly m_PropertyInfo As PropertyInfo
    Private ReadOnly m_Metadata As SettingsAttribute

    Public Sub New(ByVal targetObject As Object, ByVal propertyInfo As PropertyInfo, ByVal metadata As SettingsAttribute)
        m_TargetObject = targetObject
        m_PropertyInfo = PropertyInfo
        m_Metadata = metadata
    End Sub

    Public ReadOnly Property SettingName() As String
        Get
            Return m_PropertyInfo.Name
        End Get
    End Property

    Public ReadOnly Property Metadata() As SettingsAttribute
        Get
            Return m_Metadata
        End Get
    End Property

    Public Property Value() As Object
        Get
            Return m_PropertyInfo.GetValue(m_TargetObject, Nothing)
        End Get
        Set(ByVal value As Object)
            m_PropertyInfo.SetValue(m_TargetObject, value, Nothing)
            OnPropertyChanged("IsDefault")
        End Set
    End Property

    Public ReadOnly Property IsDefault() As Boolean
        Get
            Return Object.Equals(Value, Metadata.DefaultValue)
        End Get
    End Property

    Public Function GetValueAsString() As String
        Return ValueToString(Value)
    End Function
    Public Function GetDefaultValueAsString() As String
        Return ValueToString(Metadata.DefaultValue)
    End Function

    Private Function ValueToString(ByVal val As Object) As String
        Select Case Metadata.Type
            Case SettingType.Directory, SettingType.ShortText, SettingType.File, SettingType.LongText
                Return CType(val, String)

            Case SettingType.Int
                Return CType(val, Integer).ToString("D")

            Case SettingType.Float
                Return CType(val, Double).ToString("R")

            Case Else
                Throw New Exception(String.Format("Unknown setting type: {0}.", Metadata.Type))
        End Select
    End Function

    Public Sub SetValueAsString(ByVal s As String)
        Select Case Metadata.Type
            Case SettingType.Directory, SettingType.ShortText, SettingType.File, SettingType.LongText
                Value = s

            Case SettingType.Int
                Value = Integer.Parse(s)

            Case SettingType.Float
                Value = Double.Parse(s)

            Case Else
                Throw New Exception(String.Format("Unknown setting type: {0}.", Metadata.Type))
        End Select
    End Sub

    Private Sub OnPropertyChanged(ByVal propertyName As String)
        RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propertyName))
    End Sub

    Public Event PropertyChanged(ByVal sender As Object, ByVal e As System.ComponentModel.PropertyChangedEventArgs) Implements System.ComponentModel.INotifyPropertyChanged.PropertyChanged

End Class


Public Class SettingInfoCollection
    Implements IEnumerable(Of SettingInfo)

    Private ReadOnly m_TargetObject As Object
    Private ReadOnly m_Settings As List(Of SettingInfo)

    Public Sub New(ByVal targetObject As Object, ByVal settings As IEnumerable(Of SettingInfo))
        m_TargetObject = targetObject
        m_Settings = New List(Of SettingInfo)(settings)
    End Sub

    Public Function GetSetting(ByVal name As String) As SettingInfo
        Return m_Settings.First(Function(s) s.SettingName = name)
    End Function

    Public Shared Function GetSettings(ByVal plugin As IMGPluginBase) As SettingInfoCollection
        Dim infos = From p In plugin.GetType().GetProperties(BindingFlags.Instance Or BindingFlags.Public) _
               Where p.IsDefined(Of SettingsAttribute)() _
               Select New SettingInfo(plugin, p, p.GetCustomAttribute(Of SettingsAttribute)())
        Return New SettingInfoCollection(plugin, infos)
    End Function

    Public Function GetEnumerator() As System.Collections.Generic.IEnumerator(Of SettingInfo) Implements System.Collections.Generic.IEnumerable(Of SettingInfo).GetEnumerator
        Return m_Settings.GetEnumerator()
    End Function

    Public Function GetEnumerator1() As System.Collections.IEnumerator Implements System.Collections.IEnumerable.GetEnumerator
        Return GetEnumerator()
    End Function
End Class

File: MetaGeta/MetaGeta.DataStore/Settings/SettingInfo.cs
// Copyright 2009 Will Thomas
// 
// This file is part of MetaGeta.
// 
// MetaGeta is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// MetaGeta is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with MetaGeta. If not, see <http://www.gnu.org/licenses/>.

#region

using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Linq;
using System.Reflection;
using MetaGeta.Utilities;

#endregion

namespace MetaGeta.DataStore {
    public class SettingInfo : INotifyPropertyChanged {
        private readonly SettingsAttribute m_Metadata;
        private readonly PropertyInfo m_PropertyInfo;
        private readonly object m_TargetObject;

        public SettingInfo(object targetObject, PropertyInfo propertyInfo, SettingsAttribute metadata) {
            m_TargetObject = targetObject;
            m_PropertyInfo = propertyInfo;
            m_Metadata = metadata;
        }

        public string SettingName {
            get { return m_PropertyInfo.Name; }
        }

        public SettingsAttribute Metadata {
            get { return m_Metadata; }
        }

        public object Value {
            get { return m_PropertyInfo.GetValue(m_TargetObject, null); }
            set {
                m_PropertyInfo.SetValue(m_TargetObject, value, null);
                OnPropertyChanged("IsDefault");
            }
        }

        public bool IsDefault {
            get { return Equals(Value, Metadata.DefaultValue); }
        }

        #region INotifyPropertyChanged Members

        public event PropertyChangedEventHandler PropertyChanged;

        #endregion

        public string GetValueAsString() {
            return ValueToString(Value);
        }

        public string GetDefaultValueAsString() {
            return ValueToString(Metadata.DefaultValue);
        }

        private string ValueToString(object val) {
            switch (Metadata.Type) {
                case SettingType.Directory:
                case SettingType.ShortText:
                case SettingType.File:
                case SettingType.LongText:
                    return (string) val;

                case SettingType.ExtensionList:
                case SettingType.DirectoryList:
                case SettingType.FileList:
                    return ((ReadOnlyCollection<string>) val).JoinToString(";");

                case SettingType.Int:
                    return ((int) val).ToString("D");

                case SettingType.Float:
                    return ((double) val).ToString("R");

                case SettingType.Date:
                    return ((DateTimeOffset) val).ToString("o");

                default:
                    throw new Exception(string.Format("Unknown setting type: {0}.", Metadata.Type));
            }
        }

        public void SetValueAsString(string s) {
            switch (Metadata.Type) {
                case SettingType.Directory:
                case SettingType.ShortText:
                case SettingType.File:
                case SettingType.LongText:
                    Value = s;
                    break;

                case SettingType.ExtensionList:
                case SettingType.DirectoryList:
                case SettingType.FileList:
                    Value = new ReadOnlyCollection<string>(s.Split(';'));
                    break;

                case SettingType.Int:
                    Value = int.Parse(s);
                    break;

                case SettingType.Float:
                    Value = double.Parse(s);
                    break;

                case SettingType.Date:
                    Value = DateTimeOffset.ParseExact(s, "o", null, System.Globalization.DateTimeStyles.RoundtripKind);
                    break;

                default:
                    throw new Exception(string.Format("Unknown setting type: {0}.", Metadata.Type));
            }
        }

        private void OnPropertyChanged(string propertyName) {
            if (PropertyChanged != null)
                PropertyChanged(this, new PropertyChangedEventArgs(propertyName));
        }
    }
}

namespace MetaGeta.DataStore {
    public class SettingInfoCollection : IEnumerable<SettingInfo> {
        private readonly List<SettingInfo> m_Settings;
        private readonly object m_TargetObject;

        public SettingInfoCollection(object targetObject, IEnumerable<SettingInfo> settings) {
            m_TargetObject = targetObject;
            m_Settings = new List<SettingInfo>(settings);
        }

        #region IEnumerable<SettingInfo> Members

        public IEnumerator<SettingInfo> GetEnumerator() {
            return m_Settings.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator() {
            return GetEnumerator();
        }

        #endregion

        public SettingInfo GetSetting(string name) {
            return m_Settings.First(s => s.SettingName == name);
        }

        public static SettingInfoCollection GetSettings(IMGPluginBase plugin) {
            IEnumerable<SettingInfo> infos = from p in plugin.GetType().GetProperties(BindingFlags.Instance | BindingFlags.Public)
                                             where p.IsDefined<SettingsAttribute>()
                                             select new SettingInfo(plugin, p, p.GetCustomAttribute<SettingsAttribute>());

            return new SettingInfoCollection(plugin, infos);
        }
    }
}